src/data_device_manager/data_offer.rs

```diff
@@ -0,0 +1,579 @@
+        }
+    }
+
+    pub fn receive(&self, mime_type: String) -> Result<ReadPipe, DataOfferError> {
+        let inner = match self {
+            DataDeviceOffer::Drag(o) => o.inner(),
+            DataDeviceOffer::Selection(o) => o.inner(),
+            DataDeviceOffer::Undetermined(_) => return Err(DataOfferError::InvalidReceive), // error?
+        };
+
+        if let Some(o) = inner {
+            receive(o, mime_type).map_err(DataOfferError::Io)
+        } else {
+            Err(DataOfferError::InvalidReceive)
+        }
+    }
+
+    pub fn accept_mime_type(&mut self, serial: u32, mime_type: Option<String>) {
+        match self {
+            DataDeviceOffer::Drag(o) => o.accept_mime_type(serial, mime_type),
+            DataDeviceOffer::Selection(o) => o.accept(serial, mime_type),
+            DataDeviceOffer::Undetermined(_) => {} // error?
+        };
+    }
+
+    pub fn set_actions(&mut self, actions: DndAction, preferred_action: DndAction) {
+        match self {
+            DataDeviceOffer::Drag(o) => o.set_actions(actions, preferred_action),
+            DataDeviceOffer::Selection(_) => {}
+            DataDeviceOffer::Undetermined(o) => o.set_actions(actions, preferred_action), // error?
+        };
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct DataOfferData {
+    pub(crate) inner: Arc<Mutex<DataDeviceOffer>>,
+}
+
+impl DataOfferData {
+    pub(crate) fn push_mime_type(&self, mime_type: String) {
+        let mut inner = self.inner.lock().unwrap();
+        match inner.deref_mut() {
+            DataDeviceOffer::Drag(ref mut o) => o.mime_types.push(mime_type),
+            DataDeviceOffer::Selection(ref mut o) => o.mime_types.push(mime_type),
+            DataDeviceOffer::Undetermined(ref mut o) => o.mime_types.push(mime_type),
+        }
+    }
+
+    pub(crate) fn set_source_action(&self, action: DndAction) {
+        let mut inner = self.inner.lock().unwrap();
+        match inner.deref_mut() {
+            DataDeviceOffer::Drag(ref mut o) => o.source_actions = action,
+            DataDeviceOffer::Selection(_) => {}
+            DataDeviceOffer::Undetermined(ref mut o) => o.source_actions = action,
+        };
+    }
+
+    pub(crate) fn set_selected_action(&self, action: DndAction) {
+        let mut inner = self.inner.lock().unwrap();
+        match inner.deref_mut() {
+            DataDeviceOffer::Drag(ref mut o) => o.selected_action = action,
+            DataDeviceOffer::Selection(_) => {}    // error?
+            DataDeviceOffer::Undetermined(_) => {} // error?
+        };
+    }
+
+    pub(crate) fn to_selection_offer(&self) {
+        let mut inner = self.inner.lock().unwrap();
+        match inner.deref_mut() {
+            DataDeviceOffer::Drag(o) => {
+                *inner = DataDeviceOffer::Selection(SelectionOffer {
+                    data_offer: o.data_offer.take(),
+                    mime_types: o.mime_types.drain(..).collect(),
```

Comment:
You can use `std::mem::take()` to replace the vec with an empty vec. Which should just move the pointer instead of creating a new allocation.

The pattern is a bit awkward to deal with in Rust, but perhaps it would be better to use `inner: Arc<Mutex<Option<DataDeviceInner>>>`? So you just `.take()` the inner value, and then consume the whole `DataDeviceOffer`.